oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -117,7 +98,78 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
      * Validates access token and refreshes it if necessary.
      */
     @Override
-    public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        return tokenManager.get();
+    public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+        GrantedOAuth2AccessToken token = null;
+        if (!tokenFuture.isDone()) {
+            return tokenFuture;
+        }
+        if (!tokenFuture.isCompletedExceptionally()) {
+            token = tokenFuture.join();
+            if (isValidToken(token)) {
+                return tokenFuture;
+            }
+        }
+
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+            return this.tokenFuture;
```

Comment:
Shouldn't we check again if the future is done with valid token, although it's not highly likely? You could make use of a loop:

```java

    CompletableFuture<GrantedOAuth2AccessToken> future;
    GrantedOAuth2AccessToken token = null;
    for (;;) {
        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
        if (!tokenFuture.isDone()) {
            return tokenFuture;
        }
        if (!tokenFuture.isCompletedExceptionally()) {
            token = tokenFuture.join();
            if (isValidToken(token)) {
                return tokenFuture;
            }
        }

        future = new CompletableFuture<>();
        if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
            break;
        }
    }
```

Reply:
Thanks, that's the case that I've never thought of.
Updated the PR. Now it's more robust!